axelrod/strategies/axelrod_second.py

```diff
@@ -975,3 +975,297 @@ def strategy(self, opponent: Player) -> Action:
+    """
+
+    name = "Harrington"
+    classifier = {
+        'memory_depth': float('inf'),
+        'stochastic': True,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def __init__(self):
+        super().__init__()
+        self.mode = "Normal"
+        self.recorded_defects = 0
+        self.exit_defect_meter = 0
+        self.coops_in_first_36 = None
+        self.was_defective = False
+
+        self.prob = 0.25
+
+        self.move_history = np.zeros([4, 2])
+        self.chi_squared = None
+        self.history_row = 0
+
+        self.more_coop = 0
+        self.generous_n_turns_ago = 3
+        self.burned = False
+
+        self.defect_streak = 0
+        self.parity_streak = [0, 0]
+        self.parity_bit = 0
+        self.parity_limit = 5
+        self.parity_hits = 0
+
+    def try_return(self, to_return, lower_flags=True, inc_parity=False):
+        if lower_flags and to_return == C:
+            self.more_coop -= 1
+            self.generous_n_turns_ago += 1
+
+        if inc_parity and to_return == D:
+            self.parity_streak[self.parity_bit] += 1
+
+        return to_return
+
+    def detect_random(self, turn):
+        """
+        Calculates a modified Pearson's Chi Squared statistic on self.history,
+        and returns True (is random) if and only if the statistic is less than
+        or equal to 3.
+
+        Pearson's Chi Squared statistic = sum[ (E_i-O_i)^2 / E_i ], where O_i
+        are the observed matrix values, and E_i is calculated as number (of
+        defects) in the row times the number in the column over (total number
+        in the matrix minus 1).
+
+        We say this is modified because it differs from a usual Chi-Squared
+        test in that:
+
+        - It divides by turns minus 2 to get expected, whereas usually we'd
+          divide by matrix total.  Total equals turns minus 1, unless Defect
+          mode has been entered at any point.
+        - Terms where expected counts are less than 1 get excluded.
+        - There's a check at the beginning on the first cell of the matrix.
+        - There's a check at the beginning for the recorded number of defects.
+
+        """
+        denom = turn - 2
+
+        if self.move_history[0, 0] / denom >= 0.8:
+            return False
+        if self.recorded_defects / denom < 0.25 or self.recorded_defects / denom > 0.75:
+            return False
+
+        expected_matrix = np.outer(self.move_history.sum(axis=1), \
+                                    self.move_history.sum(axis=0))
+
+        chi_squared = 0.0
+        for i in range(4):
+            for j in range(2):
+                expct = expected_matrix[i, j] / denom
+                if expct > 1.0:
+                    chi_squared += (expct - self.move_history[i, j]) ** 2 / expct
+
+        self.chi_squared = round(chi_squared, 3) # For testing
```

Comment:
Is this round really necessary?

Reply:
I'm not sure if this is proper, but I'm using self.chi_squared just to help testing.  I'm asserting in the test file that it equals the value I'd expect; which is why I rounded.  I can remove this if you want.

Reply:
It would be better to break this out into a function that can be directly tested, otherwise let's at least have a more detailed comment like "Caching value only for testing purposes, not used otherwise" (and also in the `__init__` function. I'd prefer a separate function.

Reply:
> I'd prefer a separate function.

I agree. :+1:

Reply:
> I'd prefer a separate function.

I'm not sure what you mean by this.

Reply:
Ok cool, so the idea would be to break `def detect_random(self, turn):` in to two separate functions.

It would look *something* like this:

```
def detect_random(self, turn):
    chi_squared = calculate_chi_squared(...)
    return chi_squared <= 3:
```

Then we can test the `calculate_chi_squared` function by itself without needed to set `self.chi_squared`:

```
       actions += [(D, C)]
       self.versus_test(axelrod.Random(0.5), expected_actions=actions, seed=10)
       # The history matrix will be [[0, 2], [5, 6], [3, 6], [4, 2]]

      self.assertEqual(round(axelrod_second.calculate_chi_squared([relevant attributes]), 3), 2.395)
```

The `calculate_chi_squared` function would take relevant attributes as inputs and could just be a function in the `axelrod_second` module (not necessarily a method).

Does that make sense?

Reply:
It does...  However, I'm using the chi-squared value to test both the chi-squared calculation _and_ the history matrix.  I tried passing history_matrix as an attrs to the versus_test, but I guess assert_equal has a hard time with numpy matrices...

Reply:
> It does... However, I'm using the chi-squared value to test both the chi-squared calculation and the history matrix. I tried passing history_matrix as an attrs to the versus_test, but I guess assert_equal has a hard time with numpy matrices...

Yup, testing numpy matrix equality isn't always what you'd hope:

```python
>>> A = np.array([[1, 6], [4, 3]])
>>> A == A
array([[ True,  True],
       [ True,  True]], dtype=bool)
>>> np.array_equal(A, A)
True
```

*but* I don't think that's important here, you can still use those inputs to the chi_squared calculation? Don't worry if you don't pass them to the `versus_test`.

(A note for later/elsewhere, it would be straight forward to add the ability to test `array_equal` for numpy arrays in `versus_test`: just check the type etc...)

Reply:
Now that I know it's working, it's fine, but when I was testing, I wanted to know that the history_matrix was getting filled out correctly.  [Lot's of different matrices will give chi_squared < 3.]